common/beerocks/bcl/unit_tests/socket_event_loop_test.cpp

```diff
@@ -0,0 +1,196 @@
+            .WillOnce(Invoke([](EventType socket, LoopType *loop) -> bool {
+                loop->die();
+                return true;
+            }));
+    };
+
+    // Add the dummy socket into the event loop
+    ASSERT_TRUE(loop.add_event(timeout_socket, timeout, std::chrono::milliseconds{1}));
+
+    // Execute the event loop
+    ASSERT_GE(loop.run(), 0);
+
+    // Remove the socket from the event loop
+    ASSERT_TRUE(loop.del_event(timeout_socket));
+    close(timeout_socket->getSocketFd());
+}
+
+TEST(beerocks_socket_event_loop, repeated_timeout)
+{
+    SocketEventLoop loop(s_test_timeout);
+    StrictMock<EventHandlersMock> timeout;
+
+    // Disable the on_write handler (to prevent the loop from firing "write ready" events)
+    timeout.on_write = nullptr;
+
+    // Create a dummy socket for checking the timeout mechanism
+    auto timeout_socket = std::make_shared<Socket>(socket(AF_UNIX, SOCK_DGRAM, 0));
+    ASSERT_NE(timeout_socket, nullptr);
+
+    // Timeout event counter
+    int timeout_seq = 0;
+
+    {
+        InSequence sequence;
+
+        // Expect the timeout handler to be executed exactly 3 times
+        EXPECT_CALL(timeout, handle_timeout(timeout_socket, &loop)).Times(3);
+
+        ON_CALL(timeout, handle_timeout(timeout_socket, &loop))
+            .WillByDefault([&](EventType socket, LoopType *loop) -> bool {
+                if (++timeout_seq == 3) {
+                    loop->die();
+                }
+                return true;
+            });
+    };
+
+    // Add the dummy socket into the event loop
+    ASSERT_TRUE(loop.add_event(timeout_socket, timeout, std::chrono::milliseconds{1}));
+
+    // Execute the event loop
+    ASSERT_GE(loop.run(), 0);
+
+    // Remove the socket from the event loop
+    ASSERT_TRUE(loop.del_event(timeout_socket));
+    close(timeout_socket->getSocketFd());
+}
+
+TEST(beerocks_socket_event_loop, simple_read_write)
+{
+    SocketEventLoop loop(s_test_timeout);
+    StrictMock<EventHandlersMock> reader;
+    StrictMock<EventHandlersMock> writer;
+
+    // Disable the on_write handler (to prevent the loop from firing "write ready" events)
+    reader.on_write = nullptr;
+
+    // Create a socketpair for reader/writer simulation
+    int sv[2];
+    int rc = socketpair(AF_UNIX, SOCK_STREAM, 0, sv);
+    ASSERT_NE(-1, rc);
+
+    auto writer_socket = std::make_shared<Socket>(sv[0]);
+    auto reader_socket = std::make_shared<Socket>(sv[1]);
+
+    ASSERT_NE(writer_socket, nullptr);
+    ASSERT_NE(reader_socket, nullptr);
+
+    {
+        InSequence sequence;
+
+        EXPECT_CALL(writer, handle_write(writer_socket, &loop))
+            .WillOnce(Invoke([](EventType socket, LoopType *loop) -> bool {
+                // Send a dummy byte and remove the socket from the loop to prevent additional timeouts
+                EXPECT_EQ(1, send(socket->getSocketFd(), "X", 1, 0));
+                EXPECT_TRUE(loop->del_event(socket));
+                return true;
+            }));
```

Comment:
If we drop write events, this can be replaced simply with
```suggestion
            EXPECT_EQ(1, send(socket->getSocketFd(), "X", 1, 0));
```

Indeed, we can send something to the socket before the run() call. It will make the reader readable, so in run() we immediately call handle_read().

Combined with a global timeout, we can easily test the negative case as well:
1. Set up expectation of no read.
2. Set up expectation of timeout.
3. Run once with a short timeout.
4. Set up expectation of read.
5. Set up expectation of no timeout.
6. send()
7. Run once with short timeout.